File: VB/DiagnosticTool/DignosticController/TraceDataProcessor.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DashboardDiagnosticTool.Data
Imports DiagnosticTool.LinuxTools

Namespace DashboardDiagnosticTool

    Public Class TraceDataProcessor

        Friend Class Vertex

            Public ReadOnly Property Item As BenchmarkItem

            Public ReadOnly Property Start As DateTime

            Public ReadOnly Property Children As Dictionary(Of Integer, DashboardDiagnosticTool.Data.BenchmarkItem) = New System.Collections.Generic.Dictionary(Of Integer, DashboardDiagnosticTool.Data.BenchmarkItem)()

            Public ReadOnly Property Order As List(Of Integer) = New System.Collections.Generic.List(Of Integer)()

            Private Counts As System.Collections.Generic.Dictionary(Of Integer, Integer) = New System.Collections.Generic.Dictionary(Of Integer, Integer)()

            Public Sub New(ByVal item As DashboardDiagnosticTool.Data.BenchmarkItem, ByVal timeStamp As System.DateTime)
                Me.Item = item
                Me.Start = timeStamp
            End Sub

            Public Sub Add(ByVal child As DashboardDiagnosticTool.Data.BenchmarkItem, ByVal methodId As Integer)
                Dim real As DashboardDiagnosticTool.Data.BenchmarkItem
                If Me.Children.TryGetValue(methodId, real) Then
                    Me.Counts(methodId) += 1
                    real.Count += 1
                    real.MSecs += child.MSecs
                    real.[End] = child.[End]
                Else
                    Me.Children.Add(methodId, child)
                    Me.Counts.Add(methodId, 1)
                    Me.Order.Add(methodId)
                End If
            End Sub

            Public Sub Process(ByVal [end] As Integer, ByVal [stop] As System.DateTime)
                Me.Item.MSecs =([stop] - Me.Start).TotalMilliseconds
                Me.Item.[End] = [end]
            End Sub

            Public Function ProcessChildren() As String
                Dim id As String = Me.Item.Name
                For Each index As Integer In Me.Order
                    Me.Item.Children.Add(Me.Children(index))
                    id += "+" & index & "." & Me.Counts(index)
                Next

                Return id
            End Function
        End Class

        Private ReadOnly sessionId As Integer

        Private ReadOnly current As System.Collections.Generic.Dictionary(Of Integer, System.Collections.Generic.Stack(Of DashboardDiagnosticTool.TraceDataProcessor.Vertex)) = New System.Collections.Generic.Dictionary(Of Integer, System.Collections.Generic.Stack(Of DashboardDiagnosticTool.TraceDataProcessor.Vertex))()

        Private ReadOnly methodIds As System.Collections.Generic.Dictionary(Of String, Integer) = New System.Collections.Generic.Dictionary(Of String, Integer)()

        Public ReadOnly Property TraceItems As Dictionary(Of Integer, System.Collections.Generic.List(Of DashboardDiagnosticTool.Data.TraceItem)) = New System.Collections.Generic.Dictionary(Of Integer, System.Collections.Generic.List(Of DashboardDiagnosticTool.Data.TraceItem))()

        Private benchmarksField As System.Collections.Generic.List(Of DashboardDiagnosticTool.Data.BenchmarkItem)

        Public ReadOnly Property Benchmarks As List(Of DashboardDiagnosticTool.Data.BenchmarkItem)
            Get
                If Me.benchmarksField Is Nothing Then
                    Me.benchmarksField = Me.current.Where(Function(element) element.Value.Count = 1).[Select](Function(element)
                        Dim thread As DashboardDiagnosticTool.TraceDataProcessor.Vertex = element.Value.Pop()
                        thread.ProcessChildren()
                        Return thread.Item
                    End Function).ToList()
                End If

                Return Me.benchmarksField
            End Get
        End Property

        Public Sub New(ByVal newSessionId As Integer)
            Me.sessionId = newSessionId
        End Sub

        Private Sub AddItem(ByVal threadID As Integer, ByVal name As String, ByVal timeStamp As System.DateTime)
            Dim item = New DashboardDiagnosticTool.Data.BenchmarkItem(name, Me.sessionId, threadID)
            item.Start = Me.TraceItems(CInt((threadID))).Count
            Me.current(CInt((threadID))).Push(New DashboardDiagnosticTool.TraceDataProcessor.Vertex(item, timeStamp))
        End Sub

        Private Sub Enter(ByVal threadID As Integer, ByVal name As String, ByVal timeStamp As System.DateTime)
            Me.ProcessThread(threadID, timeStamp)
            Me.AddItem(threadID, name, timeStamp)
        End Sub

        Private Sub FixEnter(ByVal name As String, ByVal timeStamp As System.DateTime, ByVal threadId As Integer)
            If Not Me.current.ContainsKey(threadId) OrElse Not Equals(Me.current(CInt((threadId))).Peek().Item.Name, name) Then Me.Enter(threadId, name, timeStamp)
        End Sub

        Public Sub Process(ByVal traceData As DiagnosticTool.LinuxTools.DXTraceEvent)
            Dim timeStamp As System.DateTime = traceData.TimeStamp
            Dim threadId As Integer = traceData.ThreadID
            Dim methodId As Integer = Nothing
            If traceData.EventKind = DiagnosticTool.LinuxTools.EventKind.Enter Then
                Me.Enter(threadId, traceData.Name, timeStamp)
            ElseIf traceData.EventKind = DiagnosticTool.LinuxTools.EventKind.Leave Then
                Me.FixEnter(traceData.Name, timeStamp, threadId)
                Dim last As DashboardDiagnosticTool.TraceDataProcessor.Vertex = Me.current(CInt((threadId))).Pop()
                last.Process(Me.TraceItems(CInt((threadId))).Count, timeStamp)
                Dim id As String = last.ProcessChildren()
                If Not Me.methodIds.TryGetValue(id, methodId) Then
                    methodId = Me.methodIds.Count
                    Me.methodIds.Add(id, methodId)
                End If

                Me.current(CInt((threadId))).Peek().Add(last.Item, methodId)
                Me.ProcessThread(threadId, timeStamp)
            ElseIf traceData.EventKind = DiagnosticTool.LinuxTools.EventKind.Trace Then
                Me.ProcessThread(threadId, timeStamp)
                Me.TraceItems(CInt((threadId))).Add(New DashboardDiagnosticTool.Data.TraceItem(Me.sessionId, threadId) With {.EventType = traceData.EventType, .Data = traceData.Name})
            End If
        End Sub

        Private Sub ProcessThread(ByVal threadID As Integer, ByVal timeStamp As System.DateTime)
            Dim threadName As String = $"Thread{threadID}"
            Dim items As System.Collections.Generic.Stack(Of DashboardDiagnosticTool.TraceDataProcessor.Vertex) = Nothing
            If Me.current.TryGetValue(threadID, items) Then
                If items.Count = 1 Then items.Peek().Process(Me.TraceItems(CInt((threadID))).Count, timeStamp)
            Else
                Me.TraceItems.Add(threadID, New System.Collections.Generic.List(Of DashboardDiagnosticTool.Data.TraceItem)())
                Me.current.Add(threadID, New System.Collections.Generic.Stack(Of DashboardDiagnosticTool.TraceDataProcessor.Vertex)())
                Me.AddItem(threadID, threadName, timeStamp)
            End If
        End Sub
    End Class
End Namespace

File: CS/DiagnosticTool/DignosticController/TraceDataProcessor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DashboardDiagnosticTool.Data;
using DiagnosticTool.LinuxTools;

namespace DashboardDiagnosticTool {
    public class TraceDataProcessor {
        internal class Vertex {
            public BenchmarkItem Item { get; }
            public DateTime Start { get; }
            public Dictionary<int, BenchmarkItem> Children { get; } = new Dictionary<int, BenchmarkItem>();
            public List<int> Order { get; } = new List<int>();
            Dictionary<int, int> Counts = new Dictionary<int, int>();

            public Vertex(BenchmarkItem item, DateTime timeStamp) {
                Item = item;
                Start = timeStamp;
            }

            public void Add(BenchmarkItem child, int methodId) {
                BenchmarkItem real;
                if(Children.TryGetValue(methodId, out real)) {
                    Counts[methodId]++;
                    real.Count++;
                    real.MSecs += child.MSecs;
                    real.End = child.End;
                } else {
                    Children.Add(methodId, child);
                    Counts.Add(methodId, 1);
                    Order.Add(methodId);
                }
            }
            public void Process(int end, DateTime stop) {
                Item.MSecs = (stop - Start).TotalMilliseconds;
                Item.End = end;
            }
            public string ProcessChildren() {
                string id = Item.Name;
                foreach(int index in Order) {
                    Item.Children.Add(Children[index]);
                    id += "+" + index + "." + Counts[index];
                }
                return id;
            }
        }
        readonly int sessionId;
        readonly Dictionary<int, Stack<Vertex>> current = new Dictionary<int, Stack<Vertex>>();
        readonly Dictionary<string, int> methodIds = new Dictionary<string, int>();
        public Dictionary<int, List<TraceItem>> TraceItems { get; } = new Dictionary<int, List<TraceItem>>();
        List<BenchmarkItem> benchmarks;
        public List<BenchmarkItem> Benchmarks {
            get {
                if(benchmarks == null) {
                    benchmarks = current
                        .Where(element => element.Value.Count == 1)
                        .Select(element => {
                            Vertex thread = element.Value.Pop();
                            thread.ProcessChildren();
                            return thread.Item;
                        })
                        .ToList();
                }
                return benchmarks;
            }
        }
        public TraceDataProcessor(int newSessionId) {
            sessionId = newSessionId;
        }

        void AddItem(int threadID, string name, DateTime timeStamp) {
            var item = new BenchmarkItem(name, sessionId, threadID);
            item.Start = TraceItems[threadID].Count;
            current[threadID].Push(new Vertex(item, timeStamp));
        }
        void Enter(int threadID, string name, DateTime timeStamp) {
            ProcessThread(threadID, timeStamp);
            AddItem(threadID, name, timeStamp);
        }
        void FixEnter(string name, DateTime timeStamp, int threadId) {
            if(!current.ContainsKey(threadId) || current[threadId].Peek().Item.Name != name)
                Enter(threadId, name, timeStamp);
        }
        public void Process(DXTraceEvent traceData) {
            DateTime timeStamp = traceData.TimeStamp;
            int threadId = traceData.ThreadID;
            if(traceData.EventKind == EventKind.Enter) {
                Enter(threadId, traceData.Name, timeStamp);
            } else if(traceData.EventKind == EventKind.Leave) {
                FixEnter(traceData.Name, timeStamp, threadId);

                Vertex last = current[threadId].Pop();
                last.Process(TraceItems[threadId].Count, timeStamp);
                string id = last.ProcessChildren();
                if(!methodIds.TryGetValue(id, out int methodId)) {
                    methodId = methodIds.Count;
                    methodIds.Add(id, methodId);
                }
                current[threadId].Peek().Add(last.Item, methodId);
                ProcessThread(threadId, timeStamp);
            } else if(traceData.EventKind == EventKind.Trace) {
                ProcessThread(threadId, timeStamp);
                TraceItems[threadId].Add(new TraceItem(sessionId, threadId) {
                    EventType = traceData.EventType,
                    Data = traceData.Name
                });
            }
        }

        void ProcessThread(int threadID, DateTime timeStamp) {
            string threadName = $"Thread{threadID}";
            if(current.TryGetValue(threadID, out Stack<Vertex> items)) {
                if(items.Count == 1)
                    items.Peek().Process(TraceItems[threadID].Count, timeStamp);
            } else {
                TraceItems.Add(threadID, new List<TraceItem>());
                current.Add(threadID, new Stack<Vertex>());
                AddItem(threadID, threadName, timeStamp);
            }
        }
    }
}
